src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RazorBrokeredServiceBase.FactoryBase`1.cs

```diff
@@ -36,6 +36,26 @@ public async Task<object> CreateAsync(
             // Dispose the AuthorizationServiceClient since we won't be using it
             authorizationServiceClient?.Dispose();
 
+#if NET
+            // So that we can control assembly loading, we re-load ourselves in the shared Razor ALC and perform the creation there.
+            // That ensures that the service type we return is in the Razor ALC and any dependencies it needs will be handled by the
+            // Razor ALC dependency loading rather than the default ServiceHub ALC that we're in right now.
+            var assemblyInRazorAlc = RazorAssemblyLoadContext.Instance.LoadFromAssemblyName(GetType().Assembly.GetName());
+            var thisInRazorAlc = assemblyInRazorAlc.CreateInstance(GetType().FullName!)!;
```

Comment:
Forgive me if I'm missing something, but could we not cast this to a specific type and avoid the reflection below? Maybe move the initial definition of `CreateInternalAsync` to a private interface in `RazorBrokeredServiceBase`, cast to that interface here, and then just directly invoke the method?

Reply:
You can't, because the type you would cast to is from the 'wrong' ALC. As soon as we reload the assembly inside the other ALC there are effectively two different sets of types. The ones we can 'speak' are the original ones in the servicehub ALC, where the ones we want to interact with are the ones in the shared razor ALC. We don't have any way to 'speak' the razor ALC ones to cast them to. If we wanted to share an interface it would have to go into its own assembly and then we'd have to explicitly have the razor ALC defer to the servicehub for that assembly. It's doable, but not wort the effort.